DRA resourceslice controller: support publishing multiple slices

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

A driver might have more devices than can fit into a single slice.

The driver determines what each slice is meant to look like. The controller then ensures that only those slices exist. It reuses existing slices where the set of devices, as identified by their names, is the same as in some desired slice. Such slices get updated to match the desired state.

In other words, attributes and the order of devices can be changed by updating an existing slice, but adding or removing a device is done by deleting and re-creating slices.

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/kubernetes/issues/126837

#### Special notes for your reviewer:


Co-authored-by: googs1025 <googs1025@gmail.com>

The test update is partly based on
https://github.com/kubernetes/kubernetes/pull/127645.

We decided to let drivers determine the slices because that'll work better when support for partitionable devices gets added. The logic for distributing devices would become too complicated.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

This code is not built into Kubernetes itself.

/wg device-management

/cc @googs1025 @klueska

pohly committed:
DRA resourceslice controller: support publishing multiple slices

The driver determines what each slice is meant to look like. The controller
then ensures that only those slices exist. It reuses existing slices where the
set of devices, as identified by their names, is the same as in some desired
slice. Such slices get updated to match the desired state.

In other words, attributes and the order of devices can be changed by updating
an existing slice, but adding or removing a device is done by deleting and
re-creating slices.

Co-authored-by: googs1025 <googs1025@gmail.com>

The test update is partly based on
https://github.com/kubernetes/kubernetes/pull/127645.

googs1025 commented:
/cc 
Thanks for inviting this.

pohly commented:
@googs1025: I think this might be ready for an initial review. Note that compared to the test case definitions in your PR I removed the "expected total slices" and "expected changes" because those IMHO were redundant. The expected slices now reflect which were added, removed or updated.

I  intend to add the mock workqueue implementation here, to increase code coverage. It's currently at 55% even though the sync logic is mostly covered.

If you see something in the code which lacks a unit test, please let me know and I'll add one. I also take patches :grin: 

Even worse of course would be if I forgot to handle some scenario in the code! Again, let's add unit tests for that.

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go:
```diff
@@ -28,172 +32,597 @@ import (
+				Name:       ownerName,
 				UID:        test.nodeUID,
 			}
+			if test.nodeUID == "" {
```
Will our current owner be a resource object other than node?

pohly replied:
Included in https://github.com/kubernetes/kubernetes/compare/a481793493efdf4267f65c724d7aed71d6fed0e1..0d03fb153e5ef5928e1de18e1ec096c0932c734a

pohly replied:
Now I need to rebase...

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -90,7 +95,28 @@ type Pool struct {
+	// is empty. "Empty pool" is different from "no pool"
+	// because it shows that the driver is up-and-running
+	// and simply doesn't have any devices.
+	Slices []Slice
```
The new []Slice that we're adding is supposed to be associated with the resource slices, right? Each Pool can have multiple Slices, and each of these Slices corresponds to a resource slice. So, the maximum limit (128) needs to be enforced at the Driver, correct?"

pohly replied:
Correct.

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -372,66 +398,164 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+			if !apiequality.Semantic.DeepEqual(&currentSlice.Spec.Pool, &desiredPool) ||
+				!apiequality.Semantic.DeepEqual(currentSlice.Spec.NodeSelector, pool.NodeSelector) ||
+				currentSlice.Spec.AllNodes != desiredAllNodes ||
+				!apiequality.Semantic.DeepEqual(currentSlice.Spec.Devices, desiredSlice.Devices) {
```
`apiequality.Semantic.DeepEqual(currentSlice.Spec.Devices, desiredSlice.Devices)` Are there any order issues? Do we need to sort certain fields when comparing?

pohly replied:
If the order is different, the slices are not identical and need to be updated. From a semantic perspective that isn't relevant, but the controller is now specified as "creates slices exactly as intended by the driver", so order is relevant.

googs1025 replied:
got this.

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller_test.go:
```diff
@@ -28,172 +32,597 @@ import (
+					poolName1: {
+						Generation: 1,
+						Slices: []Slice{{
+							Devices: []resourceapi.Device{
```
Do we need a case that covers the scenario where the Devices list in the pool Slice contains multiple Devices?

pohly replied:
https://github.com/kubernetes/kubernetes/compare/a481793493efdf4267f65c724d7aed71d6fed0e1..0d03fb153e5ef5928e1de18e1ec096c0932c734a

googs1025 commented:
https://github.com/kubernetes/kubernetes/pull/128331 There seems to be some duplication

pohly commented:
> #128331 There seems to be some duplication

Just saw that today. I'll rebase on top of that one.

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -94,7 +95,28 @@ type Pool struct {
+
+// Slice is turned into one ResourceSlice by the controller.
+type Slice struct {
+	// Devices lists all devices which are part of the slice.
```
Do we need to indicate the 128 limit here? Because resource slices do not check the relevant limit now, the upper-level Pool needs to do the relevant verification.

pohly replied:
It's mentioned above:

         ... The driver must ensure
	// that each resulting slice is valid. See the API
	// definition for details, in particular the limit on
	// the number of devices.
	
I don't want to repeat here what "valid" means because such documentation can easily become stale.

klueska replied:
I haven't gotten deep enough in the code et to see, but I'm assuming that attempting to write a slice that has more then 128 devices will return an error, correct? I.e we will attempt to write it to the API server, it will fail validation, and then `PublishResources()` will fail, correct?

pohly committed:
DRA: validate set of devices in a pool before using the pool

The ResourceSlice controller (theoretically) might end up creating too many
slices if it syncs again before its informer cache was updated. This could
cause the scheduler to allocate a device from a duplicated slice. They should
be identical, but its still better to fail and wait until the controller
removes the redundant slice.

pohly commented:
I converted testing to use the mock work queue. That and modified test cases brings code overage up to 76%.

While thinking about repeatedly syncing the pool I noticed a potential issue with creating too many slices - see second commit. I think it is unlikely and might not even cause any problems, but I added some additional sanity checking to the allocator anyway.

As far as I am concerned, this PR is now complete.

pohly commented:
/retest

Unrelated unit test flake.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,66 +431,183 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+			// nothing because of the duplicated device IDs.
+			//
+			// To avoid creating a second slice, we would have to use a
+			// https://pkg.go.dev/k8s.io/client-go/tools/cache#MutationCache.
```
One unit test failed because of this race:
```
    resourceslicecontroller.go:526: I1025 22:13:35.488207] Bumped generation by one poolName="pool" generation=2
    resourceslicecontroller.go:544: I1025 22:13:35.488351] Updating existing resource slice poolName="pool" slice="resource-slice-2"
    resourceslicecontroller.go:544: I1025 22:13:35.488591] Updating existing resource slice poolName="pool" slice="resource-slice-3"
    resourceslicecontroller.go:298: I1025 22:13:35.488812] ResourceSlice update slice="resource-slice-2"
    resourceslicecontroller.go:298: I1025 22:13:35.488936] ResourceSlice update slice="resource-slice-3"
    resourceslicecontroller.go:544: I1025 22:13:35.488992] Updating existing resource slice poolName="pool" slice="resource-slice-1"
    resourceslicecontroller.go:434: I1025 22:13:35.489234] Existing slices poolName="pool" obsolete=["resource-slice-1"] current=["resource-slice-2","resource-slice-3"]
```

All three existing slices where updated, but the informer cache had only been updated with two updated slices. The third one was the original one with `generation: 1`, and therefore it was considered obsolete when another sync got started because of the update events.

While the end result was okay, the race caused more API calls - not what we want!

So we need the MutationCache after all.

googs1025 replied:
This doesn't seem to happen every time. I tested it locally and it seems to happen in some cases at irregular intervals.

pohly replied:
It's a classic race condition... I had tested with `go test -count=10` locally without hitting it, but that doesn't mean that there's no race :hot_face:

pohly replied:
This should now be complete, including an E2E test.

googs1025 replied:
I tested it locally several times and it seems that there is no longer a race error by using `cache.NewIntegerResourceVersionMutationCache` . 😄

pohly committed:
DRA resourceslice controller: use preconditions for Delete

It's better to verify UID and ResourceVersion of the ResourceSlice that we want
to delete. If anything changed, the decision to remove it might not apply
anymore and we need to check again.

pohly committed:
DRA resource slice controller: add E2E test

This test covers creating and deleting 100 large ResourceSlices. It is strict
about using the minimum number of calls.

The test also verifies that creating large slices works.

pohly committed:
DRA resource slice controller: use MutationCache to avoid race

This avoids the problem of creating an additional slice when the one from the
previous sync is not in the informer cache yet. It also avoids false
attempts to delete slices which were updated in the previous sync. Such
attempts would fail the ResourceVersion precondition check, but would
still cause work for the apiserver.

pohly committed:
DRA resource claim controller: add statistics

This is primarily for testing. Proper metrics might be useful, but can still be
added later.

pohly commented:
/test pull-kubernetes-unit  pull-kubernetes-kind-dra

Retesting to ensure that there are no flakes. It passed.

googs1025 commented on staging/src/k8s.io/dynamic-resource-allocation/structured/pools.go:
```diff
@@ -101,17 +104,32 @@ func addSlice(pools map[PoolID]*Pool, slice *resourceapi.ResourceSlice) {
 	pool.Slices = append(pool.Slices, slice)
 }
 
+func poolIsInvalid(pool *Pool) (bool, string) {
```
Is the only scenario where the pool is invalid when the device name is duplicated? Are there other scenarios to consider? Length limit?

pohly replied:
The length limit is enforced by the apiserver, so that can't occur. What the apiserver cannot validate are constraints across multiple objects, therefore the combined pool might be invalid.

I can't think of any other aspect that is worth checking here. The "multiple devices" one was something that came out of the observation that a consumer might see such duplicated slices if the producer malfunctions.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -94,7 +109,28 @@ type Pool struct {
+	// pool that they normally manage, even if that slice
+	// is empty. "Empty pool" is different from "no pool"
+	// because it shows that the driver is up-and-running
+	// and simply doesn't have any devices.
```
@klueska: you wanted to check whether this "should publish" is sufficient or whether we should enforce this by creating an empty slice when there is none.

klueska replied:
I have a feeling driver authors will get this wrong as it's not very intuitive to *have* to create an empty slice if you aren't actually publishing any devices.

pohly replied:
That mirrors my own thinking, so let's leave it as-is for now.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -222,7 +313,7 @@ func (c *Controller) initInformer(ctx context.Context) error {
 		options.FieldSelector = selector.String()
 	})
-	c.sliceStore = informer.GetIndexer()
+	c.sliceStore = cache.NewIntegerResourceVersionMutationCache(informer.GetStore(), informer.GetIndexer(), c.mutationCacheTTL, true /* includeAdds */)
```
Another improvement idea: in the event handlers below, add pools with a delay (`AddAfter` instead of `Add`). This doesn't matter anymore when we add slices because of the mutation cache, but if we delete slices, it might reduce the number of redundant delete calls because the sync should run after the informer cache is up-to-date again and thus all executed deletes are known.

pohly replied:
Implemented.

pohly replied:
Before:
```
  I1029 15:00:33.588378  313236 resourceslicecontroller.go:693] "Deleting obsolete resource slice" poolName="network-attached" slice="dra-9321-zlxmg" deleteOptions={"preconditions":{"uid":"b54363dd-4cc7-4518-83e5-caae96907e8d","resourceVersion":"929"}}
  I1029 15:00:33.596386  313236 resourceslicecontroller.go:366] "ResourceSlice delete" slice="dra-9321-zkch8"
  I1029 15:00:33.637719  313236 resourceslicecontroller.go:489] "Existing slices" poolName="network-attached" obsolete=["dra-9321-zlxmg"] current=["dra-9321-pbssq"]
  I1029 15:00:33.637777  313236 resourceslicecontroller.go:523] "Matched existing slice" poolName="network-attached" slice="dra-9321-pbssq" matchIndex=0
  I1029 15:00:33.637831  313236 resourceslicecontroller.go:561] "Completed comparison" poolName="network-attached" numObsolete=1 numMatchedSlices=1 numChangedMatchedSlices=0 numNewSlices=0
  I1029 15:00:33.637862  313236 resourceslicecontroller.go:577] "Kept generation because at most one update API call is necessary" poolName="network-attached" generation=2
  I1029 15:00:33.637900  313236 resourceslicecontroller.go:693] "Deleting obsolete resource slice" poolName="network-attached" slice="dra-9321-zlxmg" deleteOptions={"preconditions":{"uid":"b54363dd-4cc7-4518-83e5-caae96907e8d","resourceVersion":"929"}}
  I1029 15:00:33.645496  313236 resourceslicecontroller.go:366] "ResourceSlice delete" slice="dra-9321-zlxmg"
  I1029 15:00:33.666990  313236 resourceslicecontroller.go:699] "Resource slice was already deleted earlier" poolName="network-attached" slice="dra-9321-zlxmg"
```

After:
```
  I1029 15:07:47.792780  321917 resourceslicecontroller.go:561] "Completed comparison" poolName="network-attached" numObsolete=100 numMatchedSlices=0 numChangedMatchedSlices=0 numNewSlices=1
  I1029 15:07:47.792866  321917 resourceslicecontroller.go:581] "Bumped generation by one" poolName="network-attached" generation=2
  I1029 15:07:47.792911  321917 resourceslicecontroller.go:657] "Creating new resource slice" poolName="network-attached"
  I1029 15:07:47.795519  321917 resourceslicecontroller.go:338] "ResourceSlice add" slice="dra-8280-vh92n"
...
  I1029 15:07:50.476812  321917 resourceslicecontroller.go:366] "ResourceSlice delete" slice="dra-8280-zg8w8"
  I1029 15:07:57.796962  321917 resourceslicecontroller.go:489] "Existing slices" poolName="network-attached" obsolete=[] current=["dra-8280-vh92n"]
```

There's only a ~7 second delay between that last informer event and the next sync. That's because the workqueue implementation only shortens the delay for an already pending item, but does not lengthen it. Therefore our 30 second delay counts down from first add event at `15:07:47.795519`.

 We would like to reset the timer for each incoming event, but it's still better than no delay.

pohly committed:
DRA resourceslice controller: delay sync

When deleting a bunch of slices, the delete events queue the pool while it is
being synced. It then got synced again immediately, while the deleted slices
were still being removed from the informer cache. The obsolete slice in the
cache caused the controller to delete it again, which fails with a "not
found". That error is ignored, but this still caused extra API calls.

Now syncing gets delayed with a configuration duration (default: 30 seconds) so
the informer cache is more likely to be up-to-date when the pool gets synced
again.

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -134,15 +168,51 @@ func StartController(ctx context.Context, kubeClient kubernetes.Interface, drive
 		defer logger.V(3).Info("Stopping")
 		c.run(ctx)
 	}()
-
-	// Sync each pool once.
-	for poolName := range resources.Pools {
-		c.queue.Add(poolName)
-	}
-
```
Why is this no longer needed?

klueska replied:
Edit: looks like this was just moved here:
https://github.com/kubernetes/kubernetes/pull/128275/files#diff-af98706b50db43cf2d25a2e2f986e2a8dc7e3d624d16e7197afe67e9e20bb8a7R299-R304

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -134,15 +168,51 @@ func StartController(ctx context.Context, kubeClient kubernetes.Interface, drive
+	// a driver uninstall because garbage collection won't work.
+	Owner *Owner
+
+	// This is the initial desired set of slices. As with
+	// [Controller.Update], the controller takes ownership of the resources
+	// instance.  The content must not get modified by the caller.
```
Is there a particular reason not to do a `DeepCopy()` in this controller to prevent drivers from accidentally modifying this object / remove the need for them to do the `DeepCopy()` themselves? I know I got this wrong the first time I implemented calls to this API and I'm sure I won't be the last.

pohly replied:
Just the cost. If the caller creates a new instance, then the DeepCopy would be redundant.

But you are right, it's probably "better safe than sorry". I'll add a DeepCopy here.

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -155,7 +225,8 @@ func (c *Controller) Stop() {
 // Update sets the new desired state of the resource information.
 //
 // The controller takes over ownership, so these resources must
-// not get modified after this method returns.
+// not get modified after this method returns. [DriverResources.DeepCopy]
+// can be used by the caller to clone some existing instance.
```
Likewise here -- why can't we / don't want to just do the `DeepyCopy()` internally?

pohly replied:
https://github.com/kubernetes/kubernetes/pull/128275#discussion_r1822302950 - will add it.

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -205,10 +311,10 @@ func (c *Controller) initInformer(ctx context.Context) error {
 		resourceapi.ResourceSliceSelectorNodeName: "",
 	}
-	if c.owner.APIVersion == "v1" && c.owner.Kind == "Node" {
+	if c.owner != nil && c.owner.APIVersion == "v1" && c.owner.Kind == "Node" {
```
Does this mean that I can now omit passing an `owner` if it doesn't make much sense (e.g. for non-node-local resources)?

pohly replied:
Yes.

klueska commented on staging/src/k8s.io/dynamic-resource-allocation/structured/pools.go:
```diff
@@ -101,17 +104,32 @@ func addSlice(pools map[PoolID]*Pool, slice *resourceapi.ResourceSlice) {
 	PoolID
-	IsIncomplete bool
-	Slices       []*resourceapi.ResourceSlice
+	IsIncomplete  bool
+	IsInvalid     bool
+	InvalidReason string
```
Do we really need IsIncomplete and IsInvalid? Can IsIncomplete become a reason for being Invalid (with an InvalidReason of Incomplete).

pohly committed:
DRA resourceslice controller: do DeepCopy for driver resources

The reason for the previous behavior was unnecessary performance overhead that
occurs when the caller already provided a "fresh" copy and doesn't touch it
afterwards.

But this is something that DRA driver developers can easily get wrong, so it's
better to be safe than sorry.

klueska commented:
I didn't do a full review of all of the core logic, but from an API standpoint and usability standpoint this is a significant improvement over what we had previously.

/approve

klueska reviewed (commented):
Mostly superficial comments

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,92 +483,240 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+		for _, currentSlice := range currentSlices {
+			matched := false
+			for i := range pool.Slices {
+				desiredSlice := &pool.Slices[i]
```
Changed.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,92 +483,240 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+		// a match or obsolete.
+		currentSliceForDesiredSlice := make(map[int]*resourceapi.ResourceSlice, len(pool.Slices))
+		for _, currentSlice := range currentSlices {
+			matched := false
```
`i` is out-of-scope after the for loop, so `currentSliceForDesiredSlice[i] == nil` wouldn't compile. I also find that a bit indirect and less readable.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,92 +483,240 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+				logger.V(5).Info("Need to update slice", "slice", klog.KObj(currentSlice), "matchIndex", i)
+			}
+		}
+		numMatchedSlices := len(currentSliceForDesiredSlice)
+		numNewSlices := resourceSliceCount - numMatchedSlices
```
Changed.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,92 +483,240 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+			"numNewSlices", numNewSlices,
+		)
+
+		bumpedGeneration := false
+		switch {
+		case pool.Generation > generation:
+			// Bump up the generation if the driver asked for it, or
+			// start with a non-zero generation.
 			generation = pool.Generation
+			bumpedGeneration = true
```
I'm not entirely sure, but I couldn't rule it out completely either. Perhaps if it has some knowledge that the controller doesn't have?

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/resourceslice/resourceslicecontroller.go:
```diff
@@ -381,92 +483,240 @@ func (c *Controller) syncPool(ctx context.Context, poolName string) error {
+				// Was handled above through an update.
+				continue
+			}
+			desiredSlice := &pool.Slices[i]
```
We don't need it - removed everywhere.

klueska commented:
/lgtm
